grant-project/zcash-grant-system
# Better validation

Closes #394, closes #400, closes #388, and includes #395.

### What This Does

* Adds more frontend validation, with helpful error messages
* Adds more endpoint validation, and convert existing validation to use webargs' `validate.Length` wherever possible
* Adds the env var `PROPOSAL_TARGET_MAX` to backend and frontend to allow for a configurable maximum amount. Set to 10,000 by default.
* Adds `maxLength` to a bunch of inputs to help users not make truncated input
* Truncates most input fields before saving to model to prevent database errors. Most validation has lower limits than the db length, so they should be caught before, but this prevents catastrophe
    * This is not done for login info, since we would rather throw than have them truncate a password / email and not be able to login
* Standardizes and humanizes a lot of error messages
* Show draft save error if it happens. Should happen less now though!

### Steps to Test

* Try to create a user with invalid inputs, confirm you cannot
* Try to create a proposal with invalid inputs, confirm you cannot
* (Advanced) try to hit the endpoints directly with invalid info (really long characters, over the limit proposal or milestone targets), confirm you can't get a database error, only validation errors
